### ISSUE TITLE
View server does not shutdown when requested

### ISSUE THREAD

**Issue by planetf1** [odpi/egeria#3708]
I started 'cocoView1' - a view server with subject area & Dino configured. The configuration was not as I wanted, so I tried to shut it down - but got an error ie:

```
 
Stopping server cocoView1 ...
DELETE https://datalake:9443/open-metadata/admin-services/users/garygeeke/servers/cocoView1/instance
Response: 
{
    "class": "VoidResponse",
    "relatedHTTPCode": 500,
    "exceptionClassName": "org.odpi.openmetadata.adminservices.ffdc.exception.OMAGConfigurationErrorException",
    "exceptionCausedBy": "org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
    "actionDescription": "deactivateTemporarily",
    "exceptionErrorMessage": "OMAG-ADMIN-500-001 Method deactivateTemporarily for OMAG server cocoView1 returned an unexpected exception of org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException with message OMAG-MULTI-TENANT-400-002 The OMAG server cocoView1 has been requested to shutdown but the following services are still running: []",
    "exceptionErrorMessageId": "OMAG-ADMIN-500-001",
    "exceptionErrorMessageParameters": [
        "cocoView1",
        "deactivateTemporarily",
        "org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
        "OMAG-MULTI-TENANT-400-002 The OMAG server cocoView1 has been requested to shutdown but the following services are still running: []"
    ],
    "exceptionSystemAction": "The system is unable to configure the OMAG server.  No change was made to the server's configuration document.",
    "exceptionUserAction": "This is likely to be either an operational or logic error. Look for other errors.  Validate the request.  If you are stuck, raise an issue."
}
```
It reports services are running -- quite possibly this relates to the two view services that are active (SA/Dino) - however the exception is not reporting any running services - leaving the user confused.

This is probably a bug in error handling and it is likely correct not to stop the server

The current config doc for this server is:

(cannot be retrieved)

It seems as if after this failure the server still shuts down anyway, with an attempt to delete again then returning

```
Stopping server cocoView1 ...
DELETE https://datalake:9443/open-metadata/admin-services/users/garygeeke/servers/cocoView1/instance
Response: 
{
    "class": "VoidResponse",
    "relatedHTTPCode": 500,
    "exceptionClassName": "org.odpi.openmetadata.adminservices.ffdc.exception.OMAGConfigurationErrorException",
    "exceptionCausedBy": "org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
    "actionDescription": "deactivateTemporarily",
    "exceptionErrorMessage": "OMAG-ADMIN-500-001 Method deactivateTemporarily for OMAG server cocoView1 returned an unexpected exception of org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException with message OMAG-MULTI-TENANT-404-002 The OMAG Server Operational Services service is not available on OMAG Server cocoView1 to handle a request from user garygeeke",
    "exceptionErrorMessageId": "OMAG-ADMIN-500-001",
    "exceptionErrorMessageParameters": [
        "cocoView1",
        "deactivateTemporarily",
        "org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
        "OMAG-MULTI-TENANT-404-002 The OMAG Server Operational Services service is not available on OMAG Server cocoView1 to handle a request from user garygeeke"
    ],
    "exceptionSystemAction": "The system is unable to configure the OMAG server.  No change was made to the server's configuration document.",
    "exceptionUserAction": "This is likely to be either an operational or logic error. Look for other errors.  Validate the request.  If you are stuck, raise an issue."
}
 

```

This is probably as expected given the temporary deletion - so let's try a permanent delete

```
 
Stopping server cocoView1 ...
DELETE https://datalake:9443/open-metadata/admin-services/users/garygeeke/servers/cocoView1
Response: 
{
    "class": "VoidResponse",
    "relatedHTTPCode": 404,
    "exceptionClassName": "org.odpi.openmetadata.frameworks.connectors.ffdc.PropertyServerException",
    "actionDescription": "deactivatePermanently",
    "exceptionErrorMessage": "OMAG-MULTI-TENANT-404-002 The OMAG Server Operational Services service is not available on OMAG Server cocoView1 to handle a request from user garygeeke",
    "exceptionErrorMessageId": "OMAG-MULTI-TENANT-404-002",
    "exceptionErrorMessageParameters": [
        "OMAG Server Operational Services",
        "cocoView1",
        "garygeeke"
    ],
    "exceptionSystemAction": "The system is unable to process the request because the service is not available.",
    "exceptionUserAction": "Verify that the correct server is being called on the correct platform and that the requested service is configured to run there.  Once the correct environment is in place, retry the request."
}
```
This fails too as the server has no support....

I think there are some gaps in the operational management of the view service.
cc: @davidradl @mandy-chessell

**Comment by planetf1** [comment]
The config at this time was as per https://gist.github.com/180d7379faaeb6271dfb9432bf516f71

**Comment by mandy-chessell** [comment]
The test that is throwing this exception is looking in the service map.
The view services need to remove themselves from the service map when they shutdown.

**Comment by grahamwallis** [comment]
I believe that the Dino view service does remove itself during shutdown....

 public void shutdown()
    {
        platformInstanceMap.removeServiceInstanceFromPlatform(serverName, serviceName);
    }

**Comment by grahamwallis** [comment]
I see the same removal in Glossary Author as well.

**Comment by grahamwallis** [comment]
Looks like we need to step through the view server shutdown to see where things are going astray. I have not had a view server hang during shutdown - but will try to reproduce.

**Comment by planetf1** [comment]
I start up the server using the notebooks - I can see it's looking ok

ie
```
GET https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoView1/configuration

{
    "class": "OMAGServerConfigResponse",
    "relatedHTTPCode": 200,
    "omagserverConfig": {
        "class": "OMAGServerConfig",
        "versionId": "V2.0",
        "localServerId": "3a800286-fc1e-45dd-9d47-413c20ea4a97",
        "localServerName": "cocoView1",
        "localServerType": "View Server",
        "organizationName": "Coco Pharmaceuticals",
        "localServerURL": "https://localhost:9444",
        "localServerUserId": "cocoView1npa",
        "localServerPassword": "cocoView1passw0rd",
        "maxPageSize": 1000,
        "serverSecurityConnection": {
            "class": "Connection",
            "headerVersion": 0,
            "connectorType": {
                "class": "ConnectorType",
                "headerVersion": 0,
                "connectorProviderClassName": "org.odpi.openmetadata.metadatasecurity.samples.CocoPharmaServerSecurityProvider"
            }
        },
        "eventBusConfig": {
            "class": "EventBusConfig",
            "topicURLRoot": "egeria.omag",
            "configurationProperties": {
                "producer": {
                    "bootstrap.servers": "localhost:9092"
                },
                "consumer": {
                    "bootstrap.servers": "localhost:9092"
                }
            },
            "additionalProperties": {
                "producer": {
                    "bootstrap.servers": "localhost:9092"
                },
                "consumer": {
                    "bootstrap.servers": "localhost:9092"
                }
            }
        },
        "viewServicesConfig": [
            {
                "class": "ViewServiceConfig",
                "viewServiceId": 5000,
                "viewServiceAdminClass": "org.odpi.openmetadata.viewservices.glossaryauthor.admin.GlossaryAuthorViewAdmin",
                "viewServiceName": "Glossary Author",
                "viewServiceFullName": "Glossary Author OMVS",
                "viewServiceURLMarker": "glossary-author",
                "viewServiceDescription": "View Service for glossary authoring.",
                "viewServiceWiki": "https://egeria.odpi.org/open-metadata-implementation/view-services/glossary-author-view/",
                "viewServiceOperationalStatus": "ENABLED",
                "omagserverPlatformRootURL": "https://localhost:9444",
                "omagserverName": "cocoMDS4"
            },
            {
                "class": "ViewServiceConfig",
                "viewServiceId": 5001,
                "viewServiceAdminClass": "org.odpi.openmetadata.viewservices.rex.admin.RexViewAdmin",
                "viewServiceName": "Repository Explorer",
                "viewServiceFullName": "Repository Explorer OMVS",
                "viewServiceURLMarker": "rex",
                "viewServiceDescription": "Explore open metadata instances.",
                "viewServiceWiki": "https://egeria.odpi.org/open-metadata-implementation/view-services/rex-view/",
                "viewServiceOperationalStatus": "ENABLED",
                "omagserverPlatformRootURL": "https://localhost:9444",
                "omagserverName": "cocoMDS4"
            },
            {
                "class": "IntegrationViewServiceConfig",
                "viewServiceId": 5002,
                "viewServiceAdminClass": "org.odpi.openmetadata.viewservices.tex.admin.TexViewAdmin",
                "viewServiceName": "Type Explorer",
                "viewServiceFullName": "Type Explorer OMVS",
                "viewServiceURLMarker": "tex",
                "viewServiceDescription": "Explore the open metadata types in a repository or cohort.",
                "viewServiceWiki": "https://egeria.odpi.org/open-metadata-implementation/view-services/tex-view/",
                "viewServiceOperationalStatus": "ENABLED",
                "resourceEndpoints": [
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Platform",
                        "description": "Core Platform",
                        "platformRootURL": "https://localhost:9443",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Platform",
                        "description": "DataLake Platform",
                        "platformRootURL": "https://localhost:9444",
                        "platformName": "DataLake"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Platform",
                        "description": "Development Platform",
                        "platformRootURL": "https://localhost:9445",
                        "platformName": "Development"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Data Lake Operations",
                        "serverName": "cocoMDS1",
                        "serverInstanceName": "cocoMDS1",
                        "platformName": "DataLake"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Governance",
                        "serverName": "cocoMDS2",
                        "serverInstanceName": "cocoMDS2",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Research",
                        "serverName": "cocoMDS3",
                        "serverInstanceName": "cocoMDS3",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Business Systems",
                        "serverName": "cocoMDS5",
                        "serverInstanceName": "cocoMDS5",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Manufacturing",
                        "serverName": "cocoMDS6",
                        "serverInstanceName": "cocoMDS6",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Development",
                        "serverName": "cocoMDSx",
                        "serverInstanceName": "cocoMDSx",
                        "platformName": "Development"
                    }
                ]
            },
            {
                "class": "IntegrationViewServiceConfig",
                "viewServiceId": 5003,
                "viewServiceAdminClass": "org.odpi.openmetadata.viewservices.dino.admin.DinoViewAdmin",
                "viewServiceName": "Dynamic Infrastructure and Operations",
                "viewServiceFullName": "Dynamic Infrastructure and Operations OMVS",
                "viewServiceURLMarker": "dino",
                "viewServiceDescription": "Explore and operate an open metadata ecosystem.",
                "viewServiceWiki": "https://egeria.odpi.org/open-metadata-implementation/view-services/dino-view/",
                "viewServiceOperationalStatus": "ENABLED",
                "resourceEndpoints": [
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Platform",
                        "description": "Core Platform",
                        "platformRootURL": "https://localhost:9443",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Platform",
                        "description": "DataLake Platform",
                        "platformRootURL": "https://localhost:9444",
                        "platformName": "DataLake"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Platform",
                        "description": "Development Platform",
                        "platformRootURL": "https://localhost:9445",
                        "platformName": "Development"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Data Lake Operations",
                        "serverName": "cocoMDS1",
                        "serverInstanceName": "cocoMDS1",
                        "platformName": "DataLake"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Governance",
                        "serverName": "cocoMDS2",
                        "serverInstanceName": "cocoMDS2",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Research",
                        "serverName": "cocoMDS3",
                        "serverInstanceName": "cocoMDS3",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Data Lake Users",
                        "serverName": "cocoMDS4",
                        "serverInstanceName": "cocoMDS4",
                        "platformName": "DataLake"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Business Systems",
                        "serverName": "cocoMDS5",
                        "serverInstanceName": "cocoMDS5",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Manufacturing",
                        "serverName": "cocoMDS6",
                        "serverInstanceName": "cocoMDS6",
                        "platformName": "Core"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "Development",
                        "serverName": "cocoMDSx",
                        "serverInstanceName": "cocoMDSx",
                        "platformName": "Development"
                    },
                    {
                        "class": "ResourceEndpointConfig",
                        "resourceCategory": "Server",
                        "description": "View Server",
                        "serverName": "cocoView1",
                        "serverInstanceName": "cocoView1",
                        "platformName": "DataLake"
                    }
                ]
            }
        ],
        "repositoryServicesConfig": {
            "class": "RepositoryServicesConfig",
            "auditLogConnections": [
                {
                    "class": "Connection",
                    "headerVersion": 0,
                    "displayName": "Console",
                    "connectorType": {
                        "class": "ConnectorType",
                        "headerVersion": 0,
                        "type": {
                            "class": "ElementType",
                            "headerVersion": 0,
                            "elementOrigin": "LOCAL_COHORT",
                            "elementVersion": 0,
                            "elementTypeId": "954421eb-33a6-462d-a8ca-b5709a1bd0d4",
                            "elementTypeName": "ConnectorType",
                            "elementTypeVersion": 1,
                            "elementTypeDescription": "A set of properties describing a type of connector."
                        },
                        "guid": "4afac741-3dcc-4c60-a4ca-a6dede994e3f",
                        "qualifiedName": "Console Audit Log Store Connector",
                        "displayName": "Console Audit Log Store Connector",
                        "description": "Connector supports logging of audit log messages to stdout.",
                        "connectorProviderClassName": "org.odpi.openmetadata.adapters.repositoryservices.auditlogstore.console.ConsoleAuditLogStoreProvider"
                    },
                    "configurationProperties": {
                        "supportedSeverities": [
                            "<Unknown>",
                            "Information",
                            "Event",
                            "Decision",
                            "Action",
                            "Error",
                            "Exception",
                            "Security",
                            "Startup",
                            "Shutdown",
                            "Asset",
                            "Types",
                            "Cohort"
                        ]
                    }
                }
            ]
        },
        "auditTrail": [
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for local server's URL root to https://localhost:9444.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke removed configuration for local server type name.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for local server type name to View Server.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for local server's owning organization's name to Coco Pharmaceuticals.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for local server's userId to cocoView1npa.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for local server's password to cocoView1passw0rd.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke added configuration for an Open Metadata Server Security Connector",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for default event bus.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke set up default audit log destinations.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for view service glossary-author.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for view service rex.",
            "Mon Oct 19 14:32:38 BST 2020 garygeeke updated configuration for view service tex.",
            "Mon Oct 19 14:32:39 BST 2020 garygeeke updated configuration for view service dino.",
            "Mon Oct 19 14:32:41 BST 2020 garygeeke deployed configuration for server."
        ]
    }
}
```
GET https://localhost:9444/open-metadata/platform-services/users/garygeeke/server-platform/servers/cocoView1

{
"class": "ServerStatusResponse",
"relatedHTTPCode": 200,
"serverName": "cocoView1",
"serverStartTime": "2020-10-19T13:33:38.309+00:00",
"active": true
}
```
```
GET https://localhost:9444/open-metadata/platform-services/users/garygeeke/server-platform/servers/cocoView1/services

{
"class": "ServerServicesListResponse",
"relatedHTTPCode": 200,
"serverName": "cocoView1",
"serverServicesList": [
"Open Metadata Repository Services (OMRS)",
"Glossary Author",
"Dynamic Infrastructure and Operations",
"Repository Explorer",
"OMAG Server Operational Services",
"Type Explorer"
]
}
```
So that all looks good - so let's try and shutdown the server

```
DELETE https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoView1/instance

{
"class": "VoidResponse",
"relatedHTTPCode": 500,
"exceptionClassName": "org.odpi.openmetadata.adminservices.ffdc.exception.OMAGConfigurationErrorException",
"exceptionCausedBy": "org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
"actionDescription": "deactivateTemporarily",
"exceptionErrorMessage": "OMAG-ADMIN-500-001 Method deactivateTemporarily for OMAG server cocoView1 returned an unexpected exception of org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException with message OMAG-MULTI-TENANT-400-002 The OMAG server cocoView1 has been requested to shutdown but the following services are still running: []",
"exceptionErrorMessageId": "OMAG-ADMIN-500-001",
"exceptionErrorMessageParameters": [
"cocoView1",
"deactivateTemporarily",
"org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
"OMAG-MULTI-TENANT-400-002 The OMAG server cocoView1 has been requested to shutdown but the following services are still running: []"
],
"exceptionSystemAction": "The system is unable to configure the OMAG server. No change was made to the server's configuration document.",
"exceptionUserAction": "This is likely to be either an operational or logic error. Look for other errors. Validate the request. If you are stuck, raise an issue."
}
```
Oh so it hasn't shutdown. Note it says no change to the server config document (though it doesn't refer to operational state....)

But also it doesn't state what services are running -- the list is empty

Let's check it's configuration again
```
GET https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoView1/instance/configuration

{
"class": "OMAGServerConfigResponse",
"relatedHTTPCode": 500,
"exceptionClassName": "org.odpi.openmetadata.adminservices.ffdc.exception.OMAGConfigurationErrorException",
"exceptionCausedBy": "org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
"actionDescription": "getActiveConfiguration",
"exceptionErrorMessage": "OMAG-ADMIN-500-001 Method getActiveConfiguration for OMAG server cocoView1 returned an unexpected exception of org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException with message OMAG-MULTI-TENANT-404-002 The OMAG Server Operational Services service is not available on OMAG Server cocoView1 to handle a request from user garygeeke",
"exceptionErrorMessageId": "OMAG-ADMIN-500-001",
"exceptionErrorMessageParameters": [
"cocoView1",
"getActiveConfiguration",
"org.odpi.openmetadata.commonservices.ffdc.exceptions.PropertyServerException",
"OMAG-MULTI-TENANT-404-002 The OMAG Server Operational Services service is not available on OMAG Server cocoView1 to handle a request from user garygeeke"
],
"exceptionSystemAction": "The system is unable to configure the OMAG server. No change was made to the server's configuration document.",
"exceptionUserAction": "This is likely to be either an operational or logic error. Look for other errors. Validate the request. If you are stuck, raise an issue."
}
```
So it is shutdown?
```
GET https://localhost:9444/open-metadata/platform-services/users/garygeeke/server-platform/servers/active

{
"class": "ServerListResponse",
"relatedHTTPCode": 200,
"serverList": [
"cocoMDS1",
"cocoMDS4",
"cocoView1"
]
}

```
No it isn't - still active - checking:
```
GET https://localhost:9444/open-metadata/platform-services/users/garygeeke/server-platform/servers/cocoView1/status
{
"class": "ServerStatusResponse",
"relatedHTTPCode": 200,
"serverName": "cocoView1",
"serverEndTime": "2020-10-19T13:58:55.047+00:00",
"serverHistory": [
{
"startTime": "2020-10-19T13:33:38.309+00:00",
"endTime": "2020-10-19T13:58:55.047+00:00"
}
],
"active": true
}

```
Yes ... so perhaps the omag services aren't running?
```
GET https://localhost:9444/open-metadata/platform-services/users/garygeeke/server-platform/servers/cocoView1/services

{
"class": "ServerServicesListResponse",
"relatedHTTPCode": 200,
"serverName": "cocoView1"
}

```
Seems that way - so the services to manage the server aren't running, the server cannot be deleted, but it is kinda running

Let's try the delete again:

-- same result

And a permanent delete

```
DELETE https://localhost:9444/open-metadata/admin-services/users/garygeeke/servers/cocoView1

{
    "class": "VoidResponse",
    "relatedHTTPCode": 404,
    "exceptionClassName": "org.odpi.openmetadata.frameworks.connectors.ffdc.PropertyServerException",
    "actionDescription": "deactivatePermanently",
    "exceptionErrorMessage": "OMAG-MULTI-TENANT-404-002 The OMAG Server Operational Services service is not available on OMAG Server cocoView1 to handle a request from user garygeeke",
    "exceptionErrorMessageId": "OMAG-MULTI-TENANT-404-002",
    "exceptionErrorMessageParameters": [
        "OMAG Server Operational Services",
        "cocoView1",
        "garygeeke"
    ],
    "exceptionSystemAction": "The system is unable to process the request because the service is not available.",
    "exceptionUserAction": "Verify that the correct server is being called on the correct platform and that the requested service is configured to run there.  Once the correct environment is in place, retry the request."
}
```

Hope that helps clarify... 

* The initial attempt to shutdown refuses to as it states things are running, yet it provides an empty list of what is running - perhaps because at that point they HAVE been shutdown in fact.
* The error message also implies no change has been made, which is misleading
* Ending up in a state where a server cannot be deleted should not be allowed to happen at all - so we have an omission in the platform services to allow this cleanup? Or we cannot allow the omag operational services to be shutdown -- which from the last message looks as if what has happened.

**Comment by grahamwallis** [comment]
Nice work. Thanks @planetf1. I can see why the view services don't get shut down. They've all got lovely shutdown methods but they are never called :-) I'll raise a PR for that change and reference this issue. 

I have locally tested a fix for that and it works - so the service map actually does get emptied. This will avoid the (posthumous) exception that is generated if the service map is not empty.

However, I am slightly worried that there may be other problems entangled in the above, as there appear to a few surprising results to your tests - such as the server appearing to still be active when it should have shut down. My view service shutdown change should not help with the apparent problem with the state of the operational services service.

**Comment by grahamwallis** [comment]
With regard to the latter anomaly, It seems like the platform-services have not been told (or found out) that the server has stopped. Each request to the platform-services reported that the server was still active; meanwhile any request to the server's admin-services reported that the server (loosely speaking) was not available.

**Comment by grahamwallis** [comment]
In fact the reason for the above is that when the PropertyServerException is thrown (which is during the server instances shutdown method) it means that the updates to the server instance maps are never made. So we should probably catch the PropertyServerException, clean up the server instance maps and rethrow the exception, or we should use a finally block.

**Comment by grahamwallis** [comment]
Have just added finally block to the above mentioned PR.
This seems to fix both the shutdown of the view services and proper server instance map modification even if a problem is encountered. (I tested it with view services deliberately not being shut down and it works).
So the platform-services now know what is going on and correctly report that the server has stopped.

**Comment by grahamwallis** [comment]
Fixed by #3910
